src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiRecordReader.java

```diff
@@ -0,0 +1,119 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.servicenow.source;
+
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.plugin.servicenow.source.apiclient.ServiceNowTableAPIClientImpl;
+import io.cdap.plugin.servicenow.source.apiclient.ServiceNowTableDataResponse;
+import io.cdap.plugin.servicenow.source.util.SchemaBuilder;
+import io.cdap.plugin.servicenow.source.util.ServiceNowConstants;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Record reader that reads the entire contents of a ServiceNow table.
+ */
+public class ServiceNowMultiRecordReader extends ServiceNowRecordReader {
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowMultiRecordReader.class);
+  private final ServiceNowMultiSourceConfig multiSourcePluginConf;
+
+  ServiceNowMultiRecordReader(ServiceNowMultiSourceConfig multiSourcePluginConf) {
+    super();
+    this.multiSourcePluginConf = multiSourcePluginConf;
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException {
+    try {
+      if (results == null) {
+        fetchData();
+      }
+
+      if (!iterator.hasNext()) {
+        return false;
+      }
+
+      row = iterator.next();
+
+      pos++;
+    } catch (Exception e) {
+      throw new IOException("Exception in nextKeyValue", e);
+    }
+    return true;
+  }
+
+  @Override
+  public StructuredRecord getCurrentValue() throws IOException {
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(schema);
+    recordBuilder.set(tableNameField, tableName);
+
+    try {
+      for (Schema.Field field : tableFields) {
+        String fieldName = field.getName();
+        Object fieldValue = convertToValue(fieldName, field.getSchema(), row);
+        recordBuilder.set(fieldName, fieldValue);
+      }
+    } catch (Exception e) {
+      throw new IOException("Error decoding row from table " + tableName, e);
+    }
+    return recordBuilder.build();
+  }
+
+  public void fetchData() {
+    tableName = split.getTableName();
+    tableNameField = multiSourcePluginConf.getTableNameField();
+
+    ServiceNowTableAPIClientImpl restApi = new ServiceNowTableAPIClientImpl(multiSourcePluginConf);
+
+    // Get the table data
+    results =
+      restApi.fetchTableRecords(tableName, multiSourcePluginConf.getStartDate(), multiSourcePluginConf.getEndDate(),
+        split.getOffset(), ServiceNowConstants.PAGE_SIZE);
+
+    LOG.debug("size={}", results.size());
+    if (!results.isEmpty()) {
+      fetchSchema(restApi);
+    }
+
+    iterator = results.iterator();
+  }
+
+  public void fetchSchema(ServiceNowTableAPIClientImpl restApi) {
```

Comment:
why is this method public

Reply:
Changed access modifier from `public` to `private`.  a7e25f3